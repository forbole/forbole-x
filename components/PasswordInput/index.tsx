import {
  FilledTextFieldProps,
  IconButton,
  InputAdornment,
  TextField,
  Box,
  Typography,
  useTheme,
} from '@material-ui/core'
import React from 'react'
import passwordStrength from 'check-password-strength'
import useTranslation from 'next-translate/useTranslation'
import HidePasswordIcon from '../../assets/images/icons/icon_hide.svg'
import ShowPasswordIcon from '../../assets/images/icons/icon_show.svg'
import useIconProps from '../../misc/useIconProps'

interface PasswordInputProps extends Partial<FilledTextFieldProps> {
  withSecurityLevel?: boolean
}

const PasswordInput: React.FC<PasswordInputProps> = ({ withSecurityLevel, value, ...props }) => {
  const iconProps = useIconProps()
  const theme = useTheme()
  const { t } = useTranslation('common')
  const [isShowingPassword, setIsShowingPassword] = React.useState(false)

  const passwordSecurityLevel = React.useMemo(() => {
    return value ? passwordStrength(value).id : 0
  }, [value])
  const passwordSecurityColors = [
    theme.palette.error.main,
    theme.palette.warning.main,
    theme.palette.success.main,
  ]

  return (
    <>
      <TextField
        InputProps={{
          disableUnderline: true,
          endAdornment: (
            <InputAdornment position="end">
              <IconButton onClick={() => setIsShowingPassword((prev) => !prev)}>
                {isShowingPassword ? (
                  <ShowPasswordIcon {...iconProps} />
                ) : (
                  <HidePasswordIcon {...iconProps} />
                )}
              </IconButton>
            </InputAdornment>
          ),
        }}
        fullWidth
        type={isShowingPassword ? 'text' : 'password'}
        variant="filled"
        value={value}
        {...props}
      />
      {withSecurityLevel ? (
        <Box mt={2}>
          <Box my={2} display="flex" alignItems="center">
            <Box flex={1} height={theme.spacing(0.5)} bgcolor={passwordSecurityColors[0]} />
            <Box
              flex={1}
              height={theme.spacing(0.5)}
              bgcolor={
                passwordSecurityLevel > 0 ? passwordSecurityColors[1] : theme.palette.grey[50]
              }
            />
            <Box
              flex={1}
              height={theme.spacing(0.5)}
              bgcolor={
                passwordSecurityLevel > 1 ? passwordSecurityColors[2] : theme.palette.grey[50]
              }
            />
            <Typography
              style={{
                margin: theme.spacing(0, 1),
                color: passwordSecurityColors[passwordSecurityLevel],
              }}
              variant="body2"
            >
              {t(`password security level ${passwordSecurityLevel}`)}
            </Typography>
          </Box>
          <Typography variant="body2">{t('password caption')}</Typography>
        </Box>
      ) : null}
    </>
  )
}

export default PasswordInput
