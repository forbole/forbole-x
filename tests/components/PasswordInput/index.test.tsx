import React from 'react'
import renderer from 'react-test-renderer'
import PasswordInput from '../../../components/PasswordInput'

describe('component: PasswordInput', () => {
  it('renders correctly', () => {
    const component = renderer.create(<PasswordInput />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders correctly when password is shown', () => {
    const component = renderer.create(<PasswordInput />)
    renderer.act(() => {
      component.root.findByType('button').props.onClick()
    })
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders correctly with low security level', () => {
    const component = renderer.create(<PasswordInput withSecurityLevel value="123" />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders correctly with medium security level', () => {
    const component = renderer.create(<PasswordInput withSecurityLevel value="123qweQWE" />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders correctly with high security level', () => {
    const component = renderer.create(<PasswordInput withSecurityLevel value="123qweQWE!@#" />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

afterEach(() => {
  jest.clearAllMocks()
})
