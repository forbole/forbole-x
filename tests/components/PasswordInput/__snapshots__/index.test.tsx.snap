// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: PasswordInput renders correctly 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
>
  <div
    className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="password"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            data-file-name="SvgIcon_hide"
            fill="#e0e0e0"
            height={16}
            width={16}
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`component: PasswordInput renders correctly when password is shown 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
>
  <div
    className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            data-file-name="SvgIcon_show"
            fill="#e0e0e0"
            height={16}
            width={16}
          />
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`component: PasswordInput renders correctly with high security level 1`] = `
Array [
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="password"
        value="123qweQWE!@#"
      />
      <div
        className="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              data-file-name="SvgIcon_hide"
              fill="#e0e0e0"
              height={16}
              width={16}
            />
          </span>
        </button>
      </div>
    </div>
  </div>,
  <div
    className="MuiBox-root MuiBox-root-11"
  >
    <div
      className="MuiBox-root MuiBox-root-12"
    >
      <div
        className="MuiBox-root MuiBox-root-13"
      />
      <div
        className="MuiBox-root MuiBox-root-14"
      />
      <div
        className="MuiBox-root MuiBox-root-15"
      />
      <p
        className="MuiTypography-root MuiTypography-body2"
        style={
          Object {
            "color": "#4caf50",
            "margin": "0px 8px",
          }
        }
      >
        common:password security level 2
      </p>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      common:password caption
    </p>
  </div>,
]
`;

exports[`component: PasswordInput renders correctly with low security level 1`] = `
Array [
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="password"
        value="123"
      />
      <div
        className="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              data-file-name="SvgIcon_hide"
              fill="#e0e0e0"
              height={16}
              width={16}
            />
          </span>
        </button>
      </div>
    </div>
  </div>,
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    <div
      className="MuiBox-root MuiBox-root-2"
    >
      <div
        className="MuiBox-root MuiBox-root-3"
      />
      <div
        className="MuiBox-root MuiBox-root-4"
      />
      <div
        className="MuiBox-root MuiBox-root-5"
      />
      <p
        className="MuiTypography-root MuiTypography-body2"
        style={
          Object {
            "color": "#f44336",
            "margin": "0px 8px",
          }
        }
      >
        common:password security level 0
      </p>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      common:password caption
    </p>
  </div>,
]
`;

exports[`component: PasswordInput renders correctly with medium security level 1`] = `
Array [
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="password"
        value="123qweQWE"
      />
      <div
        className="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              data-file-name="SvgIcon_hide"
              fill="#e0e0e0"
              height={16}
              width={16}
            />
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  <div
    className="MuiBox-root MuiBox-root-6"
  >
    <div
      className="MuiBox-root MuiBox-root-7"
    >
      <div
        className="MuiBox-root MuiBox-root-8"
      />
      <div
        className="MuiBox-root MuiBox-root-9"
      />
      <div
        className="MuiBox-root MuiBox-root-10"
      />
      <p
        className="MuiTypography-root MuiTypography-body2"
        style={
          Object {
            "color": "#ff9800",
            "margin": "0px 8px",
          }
        }
      >
        common:password security level 1
      </p>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      common:password caption
    </p>
  </div>,
]
`;
